File: internal/grammar/expression_test.go
package grammar

import (
	"strings"
	"testing"
)

// Cases port packages/shared/src/expression.test.ts at the parity pin; each
// group cites the it(...) block it mirrors. The TS tests ARE the grammar
// specification.

func specScope() Scope {
	return Scope{
		Context: map[string]any{
			"http":     map[string]any{"output": map[string]any{"statusCode": float64(200), "ok": true}},
			"approval": map[string]any{"output": map[string]any{"decision": "approved"}},
			"customer": map[string]any{"output": map[string]any{
				"createdAt": "2026-07-14T12:30:00Z",
				"email":     "operator@example.com",
				"message":   "payment failed: card declined",
				"tags":      []any{"priority", "billing"},
			}},
		},
		Inputs: map[string]any{"threshold": float64(10), "allowedTags": []any{"billing", "support"}},
	}
}

func evalOK(t *testing.T, expr string, scope Scope) bool {
	t.Helper()
	got, err := EvaluateExpression(expr, scope)
	if err != nil {
		t.Fatalf("%s: unexpected error %v", expr, err)
	}
	return got
}

func TestEvaluatesAllowedComparisonsAndBooleanOperators(t *testing.T) {
	// "evaluates allowed comparisons and boolean operators"
	scope := specScope()
	if !evalOK(t, "context.http.output.statusCode === 200 && context.http.output.ok === true", scope) {
		t.Fatal("strict equality chain must hold")
	}
	if !evalOK(t, "context.approval.output.decision === 'rejected' || inputs.threshold >= 10", scope) {
		t.Fatal("or-composition must hold")
	}
}

func TestRejectsArbitraryCodeExecution(t *testing.T) {
	// "rejects expressions that try to execute arbitrary code"
	result := ValidateExpression("process.exit()")
	if result.Valid || result.Code != "unsupported_token" ||
		result.Message != "Unsupported expression token: process.exit()" ||
		result.Token != "process.exit()" {
		t.Fatalf("wire-shape parity broken: %+v", result)
	}
	for _, expr := range []string{
		"context.http.output.ok; process.exit()",
		`context.customer.output.message.includes("failed")`,
		"context.customer.output.message matches /failed/",
	} {
		if ValidateExpression(expr).Valid {
			t.Fatalf("%s must be rejected", expr)
		}
	}
}

func TestStringAndCollectionOperators(t *testing.T) {
	// "supports string and collection operators without function calls"
	scope := specScope()
	cases := []struct {
		expr string
		want bool
	}{
		{"context.customer.output.message contains 'card declined'", true},
		{"context.customer.output.email startsWith 'operator@'", true},
		{"context.customer.output.message matches 'payment *: card ?eclined'", true},
		{"'billing' in context.customer.output.tags", true},
		{"context.approval.output.decision in ['approved', 'review']", true},
		{"'fraud' in inputs.allowedTags", false},
		{"context.customer.output.tags contains 'priority'", true},
		{"true in [false, true, null, 1]", true},
	}
	for _, tc := range cases {
		if got := evalOK(t, tc.expr, scope); got != tc.want {
			t.Fatalf("%s: got %v want %v", tc.expr, got, tc.want)
		}
	}
}

func TestLexicographicStringsAndNumericComparisons(t *testing.T) {
	// "compares strings lexicographically while preserving numeric comparisons"
	scope := specScope()
	cases := []struct {
		expr string
		want bool
	}{
		{"context.customer.output.createdAt >= '2026-07-01T00:00:00Z'", true},
		{"context.customer.output.createdAt < '2027-01-01T00:00:00Z'", true},
		{"'10' < '2'", true},
		{"inputs.threshold > 2", true},
		{"inputs.threshold > '2'", true},
	}
	for _, tc := range cases {
		if got := evalOK(t, tc.expr, scope); got != tc.want {
			t.Fatalf("%s: got %v want %v", tc.expr, got, tc.want)
		}
	}
}

func TestOperatorContractValidationAndRuntimeTypeDrift(t *testing.T) {
	// "validates operator contracts against empty scopes and keeps runtime
	// type drift non-fatal"
	valid := []string{
		"context.customer.output.message contains 'failed'",
		"context.customer.output.email startsWith 'operator'",
		"context.customer.output.message matches '*failed*'",
		"context.approval.output.decision in ['approved', 'rejected']",
		"context.customer.output.createdAt >= '2026-01-01'",
	}
	for _, expr := range valid {
		if !ValidateExpression(expr).Valid {
			t.Fatalf("%s must validate", expr)
		}
	}
	invalid := []string{
		"context.value in 'not-an-array'",
		"context.value startsWith 123",
		"context.value matches 123",
		"context.value > true",
		"false && context.value in 'not-an-array'",
		"true || context.value startsWith 123",
		"context.value in [context.other]",
		"context.value in [['nested']]",
	}
	for _, expr := range invalid {
		if ValidateExpression(expr).Valid {
			t.Fatalf("%s must be rejected", expr)
		}
	}

	scope := specScope()
	drift := []string{
		"context.http.output.ok > 0",
		"context.http.output.statusCode contains '20'",
		"'approved' in context.approval.output.decision",
	}
	for _, expr := range drift {
		if evalOK(t, expr, scope) {
			t.Fatalf("%s: runtime type drift must stay non-fatal false", expr)
		}
	}
}

func TestGlobBoundsAndLinearTime(t *testing.T) {
	// "bounds glob matching and handles long wildcard inputs in linear time"
	longScope := Scope{
		Context: map[string]any{"value": strings.Repeat("a", 10_000) + "z"},
		Inputs:  map[string]any{},
	}
	if !evalOK(t, "context.value matches 'a*z'", longScope) {
		t.Fatal("long wildcard match must succeed")
	}
	oversize := "context.value matches '" + strings.Repeat("*", 257) + "'"
	if ValidateExpression(oversize).Valid {
		t.Fatal("oversize pattern must fail validation")
	}
	_, err := EvaluateExpression("'value' matches '"+strings.Repeat("*", 257)+"'", specScope())
	if err == nil || !strings.Contains(err.Error(), "pattern exceeds 256 characters") {
		t.Fatalf("oversize pattern must error at runtime, got %v", err)
	}
}

func TestUndefinedVersusNullDistinction(t *testing.T) {
	// Semantics the reference inherits from JS and the port must pin: an
	// unresolved path is undefined, which loose-equals null but never
	// strict-equals it, and two unresolved paths strict-equal each other.
	scope := Scope{Context: map[string]any{"present": nil}, Inputs: map[string]any{}}
	cases := []struct {
		expr string
		want bool
	}{
		{"context.present === null", true},
		{"context.absent === null", false},
		{"context.absent == null", true},
		{"context.absent === context.otherabsent", true},
		{"context.absent !== null", true},
	}
	for _, tc := range cases {
		if got := evalOK(t, tc.expr, scope); got != tc.want {
			t.Fatalf("%s: got %v want %v", tc.expr, got, tc.want)
		}
	}
}

func TestLooseEqualityCoercions(t *testing.T) {
	// The grammar intentionally exposes JS loose equality; pin the scalar
	// coercion table the reference relies on.
	scope := Scope{
		Context: map[string]any{"count": float64(5), "flag": true, "text": "5"},
		Inputs:  map[string]any{},
	}
	cases := []struct {
		expr string
		want bool
	}{
		{"context.count == '5'", true},
		{"context.text == 5", true},
		{"context.flag == 1", true},
		{"context.flag == '1'", true},
		{"context.count != '6'", true},
		{"context.count == 'abc'", false},
	}
	for _, tc := range cases {
		if got := evalOK(t, tc.expr, scope); got != tc.want {
			t.Fatalf("%s: got %v want %v", tc.expr, got, tc.want)
		}
	}
}

func TestEmptyExpressionCodes(t *testing.T) {
	// validateExpression code mapping: empty input and empty operand.
	if got := ValidateExpression("   "); got.Code != "empty_expression" ||
		got.Message != "Expression cannot be empty" {
		t.Fatalf("empty expression mapping broken: %+v", got)
	}
	if got := ValidateExpression("context.a === "); got.Code != "empty_value" ||
		got.Message != "Expression value cannot be empty" {
		t.Fatalf("empty value mapping broken: %+v", got)
	}
}

func TestParenthesesAndNegation(t *testing.T) {
	// Grammar composition: outer-paren stripping and negation.
	scope := specScope()
	cases := []struct {
		expr string
		want bool
	}{
		{"(context.http.output.ok === true)", true},
		{"!(context.http.output.ok === false)", true},
		{"!context.http.output.ok", false},
	}
	for _, tc := range cases {
		if got := evalOK(t, tc.expr, scope); got != tc.want {
			t.Fatalf("%s: got %v want %v", tc.expr, got, tc.want)
		}
	}

	// Verified live against the reference evaluator: a parenthesized boolean
	// group composed with a further operator is OUTSIDE the grammar — parens
	// group only at the outermost level or after `!`. The port must reject
	// it with the identical token error, not quietly support more.
	_, err := EvaluateExpression("(context.http.output.statusCode === 200 || false) && !false", scope)
	if err == nil || err.Error() != "Unsupported expression token: 200 || false" {
		t.Fatalf("paren-group limitation parity broken: %v", err)
	}
}

func TestQuotedOperatorsDoNotSplit(t *testing.T) {
	// Operator characters inside quoted strings must not split the
	// expression — the reference's splitter is quote-aware.
	scope := Scope{
		Context: map[string]any{"note": "a && b || c"},
		Inputs:  map[string]any{},
	}
	if !evalOK(t, "context.note === 'a && b || c'", scope) {
		t.Fatal("quoted operators must stay literal text")
	}
	if !evalOK(t, "context.note contains ' && '", scope) {
		t.Fatal("word operators inside quotes must not split")
	}
}

func TestDomainValidatorAdapter(t *testing.T) {
	// The domain seam receives (valid, message) — wire the adapter shape.
	if valid, _ := DomainValidator("context.a === 1"); !valid {
		t.Fatal("valid expression must pass through the adapter")
	}
	valid, message := DomainValidator("process.exit()")
	if valid || message != "Unsupported expression token: process.exit()" {
		t.Fatalf("adapter must relay the reference message, got %q", message)
	}
}

File: web/src/lib/expression.test.ts
import { describe, expect, it } from 'vitest'
import {
  evaluateExpression,
  formatSimpleComparisonExpression,
  parseSimpleComparisonExpression,
  validateExpression,
} from './expression'

describe('evaluateExpression', () => {
  const scope = {
    context: {
      http: { output: { statusCode: 200, ok: true } },
      approval: { output: { decision: 'approved' } },
      customer: {
        output: {
          createdAt: '2026-07-14T12:30:00Z',
          email: 'operator@example.com',
          message: 'payment failed: card declined',
          tags: ['priority', 'billing'],
        },
      },
    },
    inputs: { threshold: 10, allowedTags: ['billing', 'support'] },
  }

  it('evaluates allowed comparisons and boolean operators', () => {
    expect(evaluateExpression("context.http.output.statusCode === 200 && context.http.output.ok === true", scope)).toBe(true)
    expect(evaluateExpression("context.approval.output.decision === 'rejected' || inputs.threshold >= 10", scope)).toBe(true)
  })

  it('evaluates parenthesized boolean groups composed with further operators', () => {
    expect(evaluateExpression("(context.http.output.statusCode === 200 || false) && !false", scope)).toBe(true)
    expect(evaluateExpression('(true || false) && true', scope)).toBe(true)
    // Grouping must actually override the ||-splits-first precedence.
    expect(evaluateExpression('(true || false) && false', scope)).toBe(false)
    expect(evaluateExpression('true || false && false', scope)).toBe(true)
    expect(evaluateExpression('!(context.http.output.ok === false || context.http.output.statusCode === 500)', scope)).toBe(true)
    expect(evaluateExpression('((context.http.output.ok === true && inputs.threshold >= 10) || (false && true))', scope)).toBe(true)
    // Parens inside quoted strings are literal text, never grouping.
    expect(evaluateExpression("'(a || b)' === '(a || b)'", scope)).toBe(true)
    expect(validateExpression("(context.http.output.statusCode === 200 || false) && !false").valid).toBe(true)
    // Static validation still visits every branch inside a paren group.
    expect(validateExpression("(true || context.value startsWith 123) && true").valid).toBe(false)
  })

  it('rejects expressions that try to execute arbitrary code', () => {
    expect(validateExpression('process.exit()')).toEqual({
      valid: false,
      message: 'Unsupported expression token: process.exit()',
      code: 'unsupported_token',
      token: 'process.exit()',
    })
    expect(validateExpression('context.http.output.ok; process.exit()').valid).toBe(false)
    expect(validateExpression('context.customer.output.message.includes("failed")').valid).toBe(false)
    expect(validateExpression('context.customer.output.message matches /failed/').valid).toBe(false)
  })

  it('supports string and collection operators without function calls', () => {
    expect(evaluateExpression("context.customer.output.message contains 'card declined'", scope)).toBe(true)
    expect(evaluateExpression("context.customer.output.email startsWith 'operator@'", scope)).toBe(true)
    expect(evaluateExpression("context.customer.output.message matches 'payment *: card ?eclined'", scope)).toBe(true)
    expect(evaluateExpression("'billing' in context.customer.output.tags", scope)).toBe(true)
    expect(evaluateExpression("context.approval.output.decision in ['approved', 'review']", scope)).toBe(true)
    expect(evaluateExpression("'fraud' in inputs.allowedTags", scope)).toBe(false)
    expect(evaluateExpression("context.customer.output.tags contains 'priority'", scope)).toBe(true)
    expect(evaluateExpression('true in [false, true, null, 1]', scope)).toBe(true)
  })

  it('compares strings lexicographically while preserving numeric comparisons', () => {
    expect(evaluateExpression("context.customer.output.createdAt >= '2026-07-01T00:00:00Z'", scope)).toBe(true)
    expect(evaluateExpression("context.customer.output.createdAt < '2027-01-01T00:00:00Z'", scope)).toBe(true)
    expect(evaluateExpression("'10' < '2'", scope)).toBe(true)
    expect(evaluateExpression("inputs.threshold > 2", scope)).toBe(true)
    expect(evaluateExpression("inputs.threshold > '2'", scope)).toBe(true)
  })

  it('validates operator contracts against empty scopes and keeps runtime type drift non-fatal', () => {
    expect(validateExpression("context.customer.output.message contains 'failed'").valid).toBe(true)
    expect(validateExpression("context.customer.output.email startsWith 'operator'").valid).toBe(true)
    expect(validateExpression("context.customer.output.message matches '*failed*'").valid).toBe(true)
    expect(validateExpression("context.approval.output.decision in ['approved', 'rejected']").valid).toBe(true)
    expect(validateExpression("context.customer.output.createdAt >= '2026-01-01'").valid).toBe(true)
    expect(validateExpression("context.value in 'not-an-array'").valid).toBe(false)
    expect(validateExpression('context.value startsWith 123').valid).toBe(false)
    expect(validateExpression('context.value matches 123').valid).toBe(false)
    expect(validateExpression('context.value > true').valid).toBe(false)
    expect(validateExpression("false && context.value in 'not-an-array'").valid).toBe(false)
    expect(validateExpression('true || context.value startsWith 123').valid).toBe(false)
    expect(validateExpression('context.value in [context.other]').valid).toBe(false)
    expect(validateExpression("context.value in [['nested']]").valid).toBe(false)

    expect(evaluateExpression('context.http.output.ok > 0', scope)).toBe(false)
    expect(evaluateExpression("context.http.output.statusCode contains '20'", scope)).toBe(false)
    expect(evaluateExpression("'approved' in context.approval.output.decision", scope)).toBe(false)
  })

  it('bounds glob matching and handles long wildcard inputs in linear time', () => {
    const longScope = { context: { value: `${'a'.repeat(10_000)}z` }, inputs: {} }
    expect(evaluateExpression("context.value matches 'a*z'", longScope)).toBe(true)
    expect(validateExpression(`context.value matches '${'*'.repeat(257)}'`).valid).toBe(false)
    expect(() => evaluateExpression(`'value' matches '${'*'.repeat(257)}'`, scope))
      .toThrow(/pattern exceeds 256 characters/)
  })

  it('refuses to compare a ReadableStream-typed context value (clear error, no silent coercion)', () => {
    // Defense-in-depth mirroring the template-side guard. If a streaming
    // HTTP node's live stream ever escapes into context, the condition
    // evaluator must NOT coerce it to a string and silently produce a
    // wrong boolean — surface a precise error instead.
    const stream = new ReadableStream<Uint8Array>({ start(c) { c.close() } })
    const scope = { context: { http: { output: { body: stream } } }, inputs: {} }
    expect(() => evaluateExpression("context.http.output.body === 'ok'", scope))
      .toThrow(/ReadableStream/)
  })

  it('round-trips one path-versus-literal comparison for guided authoring', () => {
    expect(parseSimpleComparisonExpression("context.http.output.statusCode >= 200")).toEqual({
      left: 'context.http.output.statusCode',
      operator: '>=',
      right: 200,
    })
    expect(parseSimpleComparisonExpression(
      "context.approval.output.decision in ['approved', 'review']",
    )).toBeNull()
    expect(formatSimpleComparisonExpression({
      left: 'context.approval.output.decision',
      operator: '===',
      right: "owner's review",
    })).toBe('context.approval.output.decision === "owner\'s review"')
    const windowsPath = String.raw`C:\temp`
    const formattedPath = formatSimpleComparisonExpression({
      left: 'context.input.path',
      operator: '===',
      right: windowsPath,
    })
    expect(formattedPath).toBe(String.raw`context.input.path === 'C:\temp'`)
    expect(parseSimpleComparisonExpression(formattedPath!)).toMatchObject({ right: windowsPath })
  })

  it('keeps complex or loss-prone expressions out of guided round trips', () => {
    expect(parseSimpleComparisonExpression(
      'context.http.output.ok === true && context.http.output.statusCode === 200',
    )).toBeNull()
    expect(parseSimpleComparisonExpression("'approved' in context.approval.output.decision")).toBeNull()
    expect(formatSimpleComparisonExpression({
      left: 'context.approval.output.decision',
      operator: '===',
      right: `can't "escape" both`,
    })).toBeNull()
    expect(formatSimpleComparisonExpression({
      left: 'process.env.SECRET',
      operator: '===',
      right: 'nope',
    })).toBeNull()
  })
})
